### PR TITLE
Validate container names

### LINKED ISSUE #102: directories don't work for blobs that start with '/'
silly corner case.

### LINKED ISSUE #297: Adding a new container allows duplicate name
Chose to add container put in a lower case ~ 8 character string and saved.

The message logs in the bottom pane indicated created. Is the view of containers perhaps cached

Clicking refresh logs doesn't change result

Even though the image shows the success of a /paceedu and paceedu - i also added another pacedu that shows up in the left container list

### PR DESCRIPTION
**Pull request by cicorias**
This should close #297

Validates container names with a regex.

### CODE REVIEW

**Comment by cicorias** [comment]
@sedouard I killed the old PR and re-issued as the repo name change messed me up.  but this incorporates what @felixrieseberg requested and the message in the status.

File: app/controllers/explorer.js [hunk 1]
```diff
@@ -154,6 +154,22 @@ export default Ember.Controller.extend({
...
+	 */
+	containerNameIsValid: function (name){
+		if ('$root' === name || '$logs' === name){
+			return true;
+		}
+
+		let rv = '^(([a-z\d]((-(?=[a-z\d]))|([a-z\d])){2,62}))|(\$root)|(\$logs)$';
+		let regex = new RegExp(rv);
+		var result = regex.test(name);
+		return result;
```
Comment (felixrieseberg) [inline app/controllers/explorer.js]: Good stuff! Here's a slightly shorter version:

```
    /**
     * Validates a name which sould be DNS ok
     * @param {string} name
     * @return {boolean}
     */
    containerNameIsValid: function (name) {
        let re = /^(([a-z\d]((-(?=[a-z\d]))|([a-z\d])){2,62}))|(\$root)|(\$logs)$/;
        return (name === '$root' || name === '$logs') ? true : re.test(name);
    },
```

File: app/controllers/explorer.js [hunk 2]
```diff
@@ -549,7 +565,22 @@ export default Ember.Controller.extend({
             if (!name){
                 return;
             }
-            name = name.toLowerCase();
+
+			// Todo - something to signal user that the container name was invalid
+			name = name.toLowerCase();
+
+			if (!this.containerNameIsValid(name)){
+				this.get('notifications').addNotification(Notification.create(
```
Comment (felixrieseberg) [inline app/controllers/explorer.js]: We could just open the error modal that we have for other errors (no worries, you couldn't have known). I didn't test this code, but this *should* work:

```
let name = this.get('newContainerName') ? this.get('newContainerName').toLowerCase() : '';

if (!name) {
    return;
}

if (!this.containerNameIsValid(name)) {
    this.set('application.lastError', stringResources.containerNameInvalidMessage(name));
    this.send('openErrorModal');
}

let newContainer = this.store.createRecord('container', {name: name, id: name});
```

**Comment by felixrieseberg** [comment]
Hey! Thanks so much - I apologize in advance, but I left you two more comments in the code. Those are the last ones from me, I promise :+1:

**Comment by sedouard** [comment]
LGTM I think the modal is a better UX for this.

File: app/controllers/explorer.js [hunk 3]
```diff
@@ -545,11 +555,17 @@ export default Ember.Controller.extend({
          * Create a new container
          */
         addContainerData: function () {
-            let name = this.get('newContainerName');
-            if (!name){
-                return;
-            }
-            name = name.toLowerCase();
+			let name = this.get('newContainerName') ? this.get('newContainerName').toLowerCase() : '';
```
Comment (sedouard) [inline app/controllers/explorer.js]: @cicorias Nit - But still bothersome. can you fix the tabbing here?

Comment (cicorias) [inline app/controllers/explorer.js]: ok - odd, i had thought .editorconfig said tab not space; i'll fix.

**Comment by cicorias** [comment]
.vscode/settings.json was checked in - probably shouldn't be. but i overwrote as default VSCode is tabs and 4 as length.

**Comment by felixrieseberg** [comment]
@cicorias: That makes sense to me and is small enough that we can smuggle it in here :) I'll wait for the tests and will merge then!

**Comment by cicorias** [comment]
had to fix the other spacing issue

**Comment by sedouard** [comment]
@cicorias probably my fault. the vscode thing was my doing. we originally had the `.settings` but vs code moved it, so I just kept it in the repo

**Comment by felixrieseberg** [comment]
Wait though - tabSize should be 4. The settings file just changed, and so did the spacing here?

**Comment by sedouard** [comment]
@felixrieseberg editorconfig has always been set to indent size = 2.

**Comment by cicorias** [comment]
to confirm - 4 'spaces' - not tabs?  Not sure if there's a mix in play already.  i'll fix in the morning.

**Comment by cicorias** [comment]
Tabs or Spaces

**Comment by felixrieseberg** [comment]
Whatever get's the job done - in the end, it should be four soaces to be consistent with the rest ;-)

**Comment by sedouard** [comment]
@cicorias I think the vscode settings are consistent with whats in editorconfig (which is used by sublime). Our code mostly uses 4 space indents (which is 2 indents with the current settings), but sometimes there are 2 space indents hiding in there. I've opened https://github.com/azure-storage/deco/issues/344 to track the issue since I don't think you should have to worry about this neuance in this PR.

**Pull request merged**